### PR TITLE
add models and kkbox benchmark notebooks

### PR DESCRIPTION
**Pull request by Vincent-Maladiere**

### CODE REVIEW

**Comment by Vincent-Maladiere** [comment]
First benchmarks and models :) @ogrisel @strayMat

**Comment by ogrisel** [comment]
Thanks. Here is a link to make it easier to read the rendered notebook online:

- https://nbviewer.org/github/soda-inria/survival-analysis-benchmark/blob/add_models/notebooks/kkbox_benchmark.ipynb

**Comment by ogrisel** [comment]
It's good to see results that are competitive or even beat the best model (the Cox-Time neural network) of the [PyCox paper](https://jmlr.org/papers/v20/18-424.html).

image

**Comment by ogrisel** [comment]
Please mention Table 7 with a link to the PyCox paper after you display the table of results in the notebook.

**Comment by ogrisel** [comment]
Could you maybe try to plot the Brier Scores across time for each model as they do in "Figure 4" of the paper?

image

EDIT: I guess it's what you do with the last cell with plotly but it does not render online. I will run it locally.

It would also be great to plot the average survival curve of each mode predicted on the test along with the KM estimate of the test set. I would expect well calibrated models to match the KM on average.

**Comment by ogrisel** [comment]
I cannot get the final plotly plot to render locally (at least not without re-running everything). Can you please do a screen shot and include it as a comment to this PR?

**Comment by ogrisel** [comment]
Please also record the train and test durations for each model and include them in the table.

**Comment by ogrisel** [comment]
Please also include a pure `RandomSurvivalForest` (without GridBC or Neighbor KM post-processing) on 0.1%, 1% and maybe 10% of the kkbox training set. Maybe 10% will already be too slow.

**Comment by Vincent-Maladiere** [comment]
Also, I'm really glad that the KaplanTree (DecisionTreeRegressor + Kaplan Meier on each leaves) offer competitive performances while being significantly less heavy in computation

**Comment by ogrisel** [comment]
> Also, I'm really glad that the KaplanTree (DecisionTreeRegressor + Kaplan Meier on each leaves) offer competitive performances while being significantly less heavy in computation

Note that `max_features=8` on a single decision tree is probably useless and even harmful: injecting randomness in a single tree (not a forest) is likely to decrease the quality of the fit.

**Comment by Vincent-Maladiere** [comment]
Just added XGBSE benchmarks, unsurprisingly their results are super close to ours.
https://nbviewer.org/github/soda-inria/survival-analysis-benchmark/blob/add_models/notebooks/kkbox_benchmark.ipynb

@ogrisel images are now displayed correctly on nbviewer, I guess we just needed to wait a bit :)

File: models/yasgbt.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    
+    def __init__(self, y):
+        self.y = y
+        times, km = kaplan_meier_estimator(y["event"], y["duration"])
+        # inverse_km has x: [0.01, ..., .99] and y: [0., ..., 820.]
+        self.inverse_km = StepFunction(x=km[::-1], y=times[::-1])
+        self.max_km, self.min_km = km[0], km[-1]
+    
+    def make_sample(self):
+        y = self.y
```
Comment (ogrisel) [inline models/yasgbt.py]: It would be better not to store the training set as an attribute of the estimator, instead pass `y` as an argument of the `make_sample` method. And make that method private :)

File: models/yasgbt.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        cens = CensoringDistributionEstimator().fit(y)
+        # calculate inverse probability of censoring weight at current time point t.
+        prob_cens_t = cens.predict_proba(times)
+        prob_cens_t[prob_cens_t == 0] = np.inf
+        # calculate inverse probability of censoring weights at observed time point
+        prob_cens_y = cens.predict_proba(y["duration"])
+        prob_cens_y[prob_cens_y == 0] = np.inf
+
+        sample_weights = np.where(mask_y_0, prob_cens_y, 0)
+        sample_weights = np.where(mask_y_1, prob_cens_t, sample_weights)
```
Comment (ogrisel) [inline models/yasgbt.py]: Shouldn't sample_weight be composed of inverse censoring probabilities instead?

Comment (ogrisel) [reply models/yasgbt.py]: But then I am not sure how to bound the weights for near-zero censoring probabilities. We could add a small epsilon before taking the inverse.

Could you print the min and max values of the censoring probabilities before taking the inverse?

File: models/yasgbt.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+                
+        return times.reshape(-1, 1), yc, sample_weights
+
+
+class YASGBT(BaseEstimator, SurvivalMixin):
+
+    def __init__(self, n_iter=10, learning_rate=0.01):
```
Comment (ogrisel) [inline models/yasgbt.py]: Let's expose the other important hyperparameters of the `HistGradientBoostingRegressor` model here to be able to randomized-search them for tuning the IBS on an held out validation set.

File: models/yasgbt.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        self.y = y
+        times, km = kaplan_meier_estimator(y["event"], y["duration"])
+        # inverse_km has x: [0.01, ..., .99] and y: [0., ..., 820.]
+        self.inverse_km = StepFunction(x=km[::-1], y=times[::-1])
+        self.max_km, self.min_km = km[0], km[-1]
+    
+    def make_sample(self):
+        y = self.y
+        
+        surv_probs = np.random.uniform(self.min_km, self.max_km, y.shape[0])
```
Comment (ogrisel) [inline models/yasgbt.py]: Let's use a `rng` passed as argument to `make_sample` and initialized with `rng = check_random_state(self.random_state)` at the beginning of fit to make it possible to control the reproducibility of the sampling.

File: models/yasgbt.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+class DataSampler:
+    
+    def __init__(self, y):
+        self.y = y
+        times, km = kaplan_meier_estimator(y["event"], y["duration"])
+        # inverse_km has x: [0.01, ..., .99] and y: [0., ..., 820.]
+        self.inverse_km = StepFunction(x=km[::-1], y=times[::-1])
+        self.max_km, self.min_km = km[0], km[-1]
+    
+    def make_sample(self):
```
Comment (ogrisel) [inline models/yasgbt.py:27]: Maybe we might want to introduce a `n_repetitions_per_iter` hyper parameter with a default to 5 or so and to call the code in `make_sample` several times before vstacking the generated training set variants to be used in a single boosting iteration.

File: models/yasgbt.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+        return times.reshape(-1, 1), yc, sample_weights
+
+
+class YASGBT(BaseEstimator, SurvivalMixin):
+
+    def __init__(self, n_iter=10, learning_rate=0.01):
+        self.n_iter = n_iter
+        self.learning_rate = learning_rate
+    
+    def fit(self, X, y):
```
Comment (ogrisel) [inline models/yasgbt.py]: To make it easier to debug, let's also add a `validation_data=None` fit parameter here.

When not `None` it would be used to monitor IBS on a validation set by passing a tuple `validation_data=(X_valid, y_valid)`.

File: models/yasgbt.py [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+            monotonic_cst=monotonic_cst,
+        )
+
+        data_sampler = DataSampler(y)
+        
+        for _ in tqdm(range(self.n_iter)):
+            times, yc, sample_weight = data_sampler.make_sample()
+            Xt = np.hstack([times, X])
+            hgbc.max_iter += 1
+            hgbc.fit(Xt, yc, sample_weight=sample_weight)
```
Comment (ogrisel) [inline models/yasgbt.py]: Let's also print the IBS on the training set at each boosting iteration to check convergence. Since the training set is stochastic we might expect some fluctuations be we should observe a decreasing trend, otherwise, there is probably a bug.

We could even take a single training set sample from the output of `make_sample` and keep it fixed to compute the training score without the sampling randomness.

File: models/yasgbt.py (hunk 2 above)
Comment (Vincent-Maladiere) [reply models/yasgbt.py]: This is close to what I have for every fold of the CV:
```python
min prob_cens_y: 0.0002805095479529531, max prob_cens_y: inf
min prob_cens_t: 0.0002805095479529531, max prob_cens_t: 0.9999914021959312
```

Comment (Vincent-Maladiere) [reply models/yasgbt.py]: adding a small epsilon like `1e-8 ` looks fine to me, WDYT?

File: models/yasgbt.py (hunk 5 above)
Comment (Vincent-Maladiere) [reply models/yasgbt.py:27]: I'm not sure to get the intuition behind this?

File: models/yasgbt.py (hunk 1 above)
Comment (ogrisel) [reply models/yasgbt.py]: Ignore this comment, I had forgotten that this was not a method of the estimator but a method from a helper class.

File: models/yasgbt.py [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+        
+        self.hgbc_ = hgbc
+        self.data_sampler_ = data_sampler
```
Comment (ogrisel) [inline models/yasgbt.py]: No need to store the sampler on the estimator because it is never used after calling fit, no?

File: models/yasgbt.py (hunk 3 above)
Comment (Vincent-Maladiere) [reply models/yasgbt.py]: Ok, could we achieve this by passing an `est_params` dict to `YASGBT`?

File: models/yasgbt.py (hunk 4 above)
Comment (ogrisel) [reply models/yasgbt.py]: Or this could be done in the constructor of the `DataSampler` class and stored as an attribute of that sampler instance, because the constructor is called at the beginning of `fit`.

File: models/yasgbt.py [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+from sklearn.base import BaseEstimator
+from sklearn.ensemble import HistGradientBoostingRegressor
+
+from sksurv.functions import StepFunction
+from sksurv.nonparametric import kaplan_meier_estimator, CensoringDistributionEstimator
+
+from models.survival_mixin import SurvivalMixin
+
+
+class DataSampler:
```
Comment (ogrisel) [inline models/yasgbt.py]: Let's call this `IBSTrainingSampler` to make it more explicit that this sampler implements the weighting scheme of the IBS formula.

File: models/yasgbt.py (hunk 8 above)
Comment (Vincent-Maladiere) [reply models/yasgbt.py]: It was for debugging purposes, but you're right

File: models/yasgbt.py (hunk 7 above)
Comment (ogrisel) [reply models/yasgbt.py]: Actually, scratch that: we can measure the exact IBS (without relying on the data sampler) here. It might be more costly though but it's just for the sake of debugging.

File: models/yasgbt.py [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+        prob_cens_t = cens.predict_proba(times)
+        prob_cens_t[prob_cens_t == 0] = np.inf
+        # calculate inverse probability of censoring weights at observed time point
+        prob_cens_y = cens.predict_proba(y["duration"])
+        prob_cens_y[prob_cens_y == 0] = np.inf
+
+        sample_weights = np.where(mask_y_0, prob_cens_y, 0)
+        sample_weights = np.where(mask_y_1, prob_cens_t, sample_weights)
+                
+        return times.reshape(-1, 1), yc, sample_weights
```
Comment (ogrisel) [inline models/yasgbt.py:50]: Maybe you could write a check in the notebook that uses the sampler to compute the loss stochastic IBS loss of the fitted model and check that it converges to the regular way of computing the IBS when averaging over several rounds of sampling.

File: models/yasgbt.py (hunk 3 above)
Comment (ogrisel) [reply models/yasgbt.py]: If you wish, but in scikit-learn we try to flatten the estimator parameters as much as possible to make it nicer for the users to see the list of default parameters. That leads to a bit more verbosity but it's more user friendly (e.g. to do a grid search).

File: models/yasgbt.py (hunk 7 above)
Comment (Vincent-Maladiere) [reply models/yasgbt.py]: Ok, do we prefer printing the IBS for a validation set (as you suggested in the comment above), for each training iteration, or both?

File: models/yasgbt.py (hunk 3 above)
Comment (Vincent-Maladiere) [reply models/yasgbt.py]: got it, thanks

File: models/yasgbt.py (hunk 2 above)
Comment (ogrisel) [reply models/yasgbt.py]: Ok let's do this for now.

File: models/yasgbt.py (hunk 5 above)
Comment (ogrisel) [reply models/yasgbt.py:27]: To make the fit less stochastic: the first iterations (with the largest learning rates) would be less sensitive to particular values of the sampled times.

But this would grow the memory usage by a factor of `n_repetitions_per_iter`.

File: models/yasgbt.py [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+        # calculate inverse probability of censoring weight at current time point t.
+        prob_cens_t = cens.predict_proba(times)
+        prob_cens_t[prob_cens_t == 0] = np.inf
+        # calculate inverse probability of censoring weights at observed time point
+        prob_cens_y = cens.predict_proba(y["duration"])
+        prob_cens_y[prob_cens_y == 0] = np.inf
+
+        sample_weights = np.where(mask_y_0, prob_cens_y, 0)
+        sample_weights = np.where(mask_y_1, prob_cens_t, sample_weights)
+
```
Comment (ogrisel) [inline models/yasgbt.py]: I think we should enforce the following to avoid introducing non-zero terms in the loss because of the censoring pattern:

```python
    # Ensure that all other terms (censored data points with a duration below the
    # sampled time) contribute nothing to the loss. The value of the classification
    # target y_c is undefined for those:
    remaining_terms = np.logical_not(mask_y_0) & np.logical_not(mask_y_1)
    sample_weights[remaining_terms] = 0
```

The inverse censoring probability weighting alone would be enough to zero out those terms: the ICPW should be between 1 and +inf by construction.

Comment (ogrisel) [reply models/yasgbt.py]: Actually this is indeed redundant with what is done in:

```python
        sample_weights = np.where(mask_y_0, prob_cens_y, 0)
        sample_weights = np.where(mask_y_1, prob_cens_t, sample_weights)
```

because of the right hand side `0` in the first where.

Comment (ogrisel) [reply models/yasgbt.py]: It might be interesting to add an inline comment to explain that the `y_c` terms that do not match either of `mask_y_0` and `mask_y_1` (censored data points with a duration below the sampled time) have 0 weight as they should not contribute to the time-varying BS loss.

File: models/yasgbt.py (hunk 7 above)
Comment (ogrisel) [reply models/yasgbt.py]: I would say both, at least while we are developping this estimator and tuning its default hyper params.

**Comment by Vincent-Maladiere** [comment]
Close and separate in multiple PR to simplify the conversation
